core/src/main/java/org/mule/runtime/core/execution/ModuleFlowProcessingPhase.java

```diff
@@ -134,36 +137,28 @@ public void runPhase(final ModuleFlowProcessingPhaseTemplate template, final Mes
       // TODO MULE-11167 Policies should be non blocking
       if (!enableSourcePolicies) {
         just(templateEvent)
-            .doOnNext(request -> {
-              fireNotification(messageProcessContext.getMessageSource(), request,
-                               messageProcessContext.getFlowConstruct(),
-                               MESSAGE_RECEIVED);
-            })
-            .then(request -> from(template.routeEventAsync(request)))
-            .doOnSuccess(getSuccessConsumer(messageSource, templateEvent, messageProcessContext, phaseResultNotifier, template,
-                                            terminateConsumer))
-            .doOnError(MessagingException.class, me -> {
-              if (me.getCause() instanceof SourceErrorException
-                  && sourceResponseErrorTypeMatcher.match(((SourceErrorException) me.getCause()).getErrorType())) {
-                try {
-                  handleSourceError(exceptionHandler, errorConsumer, (SourceErrorException) me.getCause());
-                  onTerminate(terminateConsumer, left(me.getEvent()));
-                } catch (Exception e) {
-                  onTerminate(terminateConsumer, right(e));
-                }
+            .doOnNext(request -> fireNotification(messageProcessContext.getMessageSource(), request,
+                                                  messageProcessContext.getFlowConstruct(),
+                                                  MESSAGE_RECEIVED))
+            .then(request -> from(template.routeEventAsync(request)).switchIfEmpty(fromCallable(() -> emptyEvent(templateEvent))))
```

Comment:
Why is this `switchIfEmpty` needed here? Is it related to this issue or a piggy-back fix.  I assume this is to handle the case where there is a filter in the flow that drop the event to ensure onError/onSuccess/onTerminate behaves consistently in this case too, but a comment to explain this would help.